src/main/ruby/core/io.rb

```diff
@@ -562,13 +562,14 @@ def self.foreach(name, separator=undefined, limit=undefined, options=undefined)
       io = File.open(name, options)
     end
 
+    each_reader = Truffle.privately { io.create_each_reader(separator, limit) }
+
     begin
-      while line = io.gets(separator, limit)
-        yield line
-      end
+      each_reader&.each { |line| yield line }
```

Comment:
Could we just pass the block argument (say `&block` here) so there is no extra block in between?

Reply:
That ought be doable. I'll look at that.